include/drivers/clock_control.h

```diff
@@ -103,9 +103,16 @@ struct clock_control_driver_api {
 
 /**
  * @brief Enable the clock of a sub-system controlled by the device
- * @param dev Pointer to the device structure for the clock controller driver
- * 	instance
- * @param sys A pointer to an opaque data representing the sub-system
+ *
+ * This function is blocking: on success, the clock is enabled and
```

Comment:
How should it behave when called from interrupt context? Do we forbid that or it should be blocking wait (with `k_cpu_idle`)?

Reply:
Yep, that is the 64 thousand dollar question.

We have a few choices which I was planning to discuss at API meeting.

Reply:
Why would it be any different from other blocking functions in this regard?

Reply:
Looks like we won't get to discuss in the API meeting. Here are some options:

1. busy-wait in ISR context, sleep in thread context
2. busy-wait no matter the context
3. fail in ISR, sleep in thread
4. ???

I'm not sure if there are other contexts we need to worry about. I don't know how (meta IRQs)[https://docs.zephyrproject.org/latest/reference/kernel/scheduling/index.html?highlight=meta%20irq#meta-irq-priorities] work, for example.

Considering that some clocks take a long time to start up and stabilize, sleeping the thread seems desirable, if a bit more complex.

I'd like to discuss in the meeting to get as much feedback as possible from people who are interested.

Reply:
@carlescufi points out this is a special case of https://github.com/zephyrproject-rtos/zephyr/issues/6184.

Reply:
we can have `k_cpu_idle()` in ISR and semaphore in thread context.

Personally, I would prefer that function works in both context. I wouldn't like to have `_isr`, rather function should internally handle it.

Reply:
Updated following discussion in the API meeting. The intent has apparently been that unless specifically marked, all driver APIs are only safe to call from threads.